src/Storages/MergeTree/CommonANNIndexes.cpp

```diff
@@ -0,0 +1,517 @@
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTSetQuery.h>
+
+#include <Storages/MergeTree/CommonANNIndexes.h>
+#include <base/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int INCORRECT_QUERY;
+}
+
+namespace ANNCondition
+{
+
+ANNCondition::ANNCondition(const SelectQueryInfo & query_info,
+                                 ContextPtr context)
+{
+    // Initialize
+    block_with_constants = KeyCondition::
+    getBlockWithConstants(query_info.query, query_info.syntax_analyzer_result, context);
+    // Build rpns for query sections
+    buildRPN(query_info);
+    // Match rpns with supported types
+    index_is_useful = matchAllRPNS();
+}
+
+bool ANNCondition::alwaysUnknownOrTrue(String metric_name) const
+{
+    if (!index_is_useful)
```

Comment:
return metric_name == ann_expr->metric_name ? !index_is_useful : true;

It is more readable, I think. If metrics are similar, return !index_is_useful